ccore/include/pyclustering/parallel/parallel.hpp

```diff
@@ -148,40 +212,36 @@ Advanced uses might use one of the define to use specific implementation of the
 @param[in] p_begin: initial iterator from that the loop starts.
 @param[in] p_end: end iterator that defines when the loop should stop `iter != p_end`.
 @param[in] p_task: body of the loop that defines actions that should be done for each element.
+@param[in] p_threads: amount of threads that are going to be used for processing (by default efficient amount of threads).
 
 */
 template <typename TypeIter, typename TypeAction>
-void parallel_for_each(const TypeIter p_begin, const TypeIter p_end, const TypeAction & p_task) {
+void parallel_for_each(const TypeIter p_begin, const TypeIter p_end, const TypeAction & p_task, const std::size_t p_threads = AMOUNT_THREADS) {
 #if defined(PARALLEL_IMPLEMENTATION_ASYNC_POOL)
-    const std::size_t step = std::distance(p_begin, p_end) / (AMOUNT_THREADS + 1);
+    const std::size_t interval_length = std::distance(p_begin, p_end);
+    const std::size_t step = std::max(interval_length / p_threads, std::size_t(1));
+
+    std::size_t amount_threads = static_cast<std::size_t>(interval_length / step);

```

Comment:
- [ ] Useless code: step was calculated via amount of threads.

Reply:
It helps to protect from out of range in case of `amount of threads < data size`.